nix/hevm-tests/smt-checker.nix

```diff
@@ -0,0 +1,383 @@
+  # $2 == hevm smt backend
+  checkWithHevm = pkgs.writeShellScript "checkWithHevm" ''
+
+    # write json file to store for later debugging
+    json=$out/jsonFiles/$(${testName} $1).json
+    ${mkdir} -p $(${echo} "''${json%/*}/")
+
+    ${solc} --combined-json=srcmap,srcmap-runtime,bin,bin-runtime,ast,metadata,storage-layout,abi $1 2> /dev/null > $json
+
+    contracts=($(${cat} $json | ${jq} .contracts | ${jq} keys | ${jq} -r '. | @sh' | ${tr} -d \'))
+
+    explore() {
+      set -x
+      hevm_output=$(${timeout} 60s ${hevm} symbolic --code "$1" --solver "$2" --json-file "$3" $4 2>&1)
+      status=$?
+      set +x
+
+      # handle timeouts
+      if [[ $status == 124 ]]; then ${echo} && ${echo} "${strings.timeout}"; fi
+
+      # handle non assertion related hevm failures
+      if [[ $status == 1 ]]; then
+        ${grep} -q '${strings.hevmAssertionViolation}' <<< "$hevm_output"
+        if [ $? == 0 ]; then return; fi
+        ${echo} "${strings.hevmCouldNotExplore}"
+      fi
+
+      # handle errors in branches
+      if [[ $status == 0 ]]; then
+        ${grep} -q '${strings.hevmErrorInBranch}' <<< "$hevm_output"
+        if [ $? == 1 ]; then return; fi
+        ${echo} "${strings.hevmCouldNotExplore}"
+      fi
+    }
+
+    for contract in "''${contracts[@]}"; do
+      ${echo}
+      ${echo} --- exploring $(basename $contract) ---
+
+      ${echo}
+      ${echo} exploring init bytecode:
+      bin=$(${jq} -r --arg c $contract -c '.contracts[$c]."bin"' $json)
+      explore "$bin" "$2" "$json" "--create"
+
+      ${echo}
+      ${echo} exploring runtime bytecode:
+      bin_runtime=$(${jq} -r --arg c $contract -c '.contracts[$c]."bin-runtime"' $json)
+      explore "$bin_runtime" "$2" "$json"
+    done
+
+    exit 0
+  '';
+
+  # Takes one Solidity file in the format of an SMTChecker test,
+  # runs `hevm symbolic` on it and compares results against the SMTChecker expectations.
+  # $1 == input file
+  # $2 == hevm smt backend
+  runSingleTest = pkgs.writeShellScript "runSingleTest" ''
+    print_test() {
+      ${echo}
+      ${echo} "*** Input File: $(${testName} $1) ***"
+      ${echo}
+      ${cat} "$1"
+    }
+
+    testName=$(${testName} $1)
+
+    ${divider}
+    ${echo} "${strings.exeucting} $(${testName} $1)"
+
+    ignoredTests=(${toString ignored})
+    if [[ " ''${ignoredTests[@]} " =~ " ''${testName} " ]]; then
+        ${echo} "${strings.ignore}"
+        exit 0
+    fi
+
+    ${grep} -q 'Error trying to invoke SMT solver.' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
+    ${grep} -q 'Assertion checker does not yet implement' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
+    ${grep} -q 'Assertion checker does not yet support' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
+    ${grep} -q 'Internal error: Expression undefined for SMT solver.' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
```

Comment:
Do you need to exit if that's the case (for the block) above? It could still report some assertion violation

Reply:
oh, I didn't realise that was the case. Removing this didn't change the numbers in the summary as the only new test executed was an SMTChecker false positive.